File: src/main/java/org/example/Borrower.java
package org.example;

import java.util.ArrayList;
import java.util.List;

public class Borrower {
    private String username;
    private String password;
    private int numBorrowedBooks;
    private List<Book> borrowedBooks;


    //constructor
    public Borrower(String username, String password) {
        this.username = username;
        this.password = password;
        this.numBorrowedBooks = 0;
        this.borrowedBooks = new ArrayList<>();
    }

    public String getUsername(){
        return username;
    }

    public String getPassword(){
        return password;
    }

    public int getNumBorrowedBooks(){
        return numBorrowedBooks;
    }

    public List<Book> getBorrowedBooks(){
        return borrowedBooks;
    }

    public void addNumBorrowedBooks(){
        numBorrowedBooks += 1;
    }

    public void addBorrowedBook(Book book){
        borrowedBooks.add(book);
    }

    public boolean atBorrowLimit(){
        return numBorrowedBooks == 3;
    }

    public void removeBorrowedBook(Book book){
        borrowedBooks.remove(book);
    }

    public void decreaseNumBorrowedBooks(){
        if (numBorrowedBooks > 0){
            numBorrowedBooks -= 1;
        }
    }
}

File: src/models/Borrower.js
//models/Borrower.js
class Borrower {
    constructor(username, password) {
        this.username = username;
        this.password = password;
        this.numBorrowedBooks = 0;
        this.borrowedBooks = []; // array of Book objects
        this.BORROW_LIMIT = 3;
    }

    // getters
    getUsername() {
        return this.username;
    }

    getPassword() {
        return this.password;
    }

    getNumBorrowedBooks() {
        return this.numBorrowedBooks;
    }

    getBorrowedBooks() {
        return this.borrowedBooks;
    }

    // actions
    addBorrowedBook(book) {
        this.borrowedBooks.push(book);
        this.numBorrowedBooks += 1;
    }

    removeBorrowedBook(book) {
        this.borrowedBooks = this.borrowedBooks.filter(b => b !== book);
        if (this.numBorrowedBooks > 0) {
            this.numBorrowedBooks -= 1;
        }
    }

    atBorrowLimit() {
        return this.numBorrowedBooks >= this.BORROW_LIMIT;
    }
}

module.exports = Borrower;